src/appleseed/foundation/image/oiioimagefilewriter.cpp

```diff
@@ -0,0 +1,515 @@
+// Visit http://appleseedhq.net/ for additional information and resources.
+//
+// This software is released under the MIT license.
+//
+// Copyright (c) 2010-2013 Francois Beaune, Jupiter Jazz Limited
+// Copyright (c) 2014-2018 Francois Beaune, The appleseedhq Organization
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+//
+
+// Interface header.
+#include "oiioimagefilewriter.h"
+
+// appleseed.foundation headers.
+#include "foundation/core/exceptions/exceptionioerror.h"
+#include "foundation/image/icanvas.h"
+#include "foundation/math/vector.h"
+#include "foundation/utility/foreach.h"
+#include "foundation/utility/iostreamop.h"
+#include "foundation/utility/string.h"
+
+// Boost headers.
+#include "boost/filesystem/path.hpp"
+
+// Standard headers.
+#include <algorithm>
+#include <memory>
+#include <stdexcept>
+
+namespace foundation
+{
+
+struct OIIOImageFileWriter::OIIOImages
+{
+    std::vector<const ICanvas*>     m_canvas;
+    std::vector<OIIO::ImageSpec>    m_spec;
+};
+
+OIIOImageFileWriter::OIIOImageFileWriter(const char* filename) :
+    m_images{ new OIIOImages{} }
+{
+    if (filename == nullptr)
+        throw ExceptionIOError("Valid filename must be provide!");
+
+    m_filename = filename;
+
+    m_writer = OIIO::ImageOutput::create(m_filename);
+    if (m_writer == nullptr)
+    {
+        const std::string msg = OpenImageIO::geterror();
+        throw ExceptionIOError(msg.c_str());
+    }
+}
+
+OIIOImageFileWriter::~OIIOImageFileWriter()
+{
+    // Destroy the ImageOutput stucture.
+    OIIO::ImageOutput::destroy(m_writer);
+
+    m_writer = nullptr;
+    m_filename = nullptr;
+
+    m_images->m_canvas.clear();
+    m_images->m_spec.clear();
+
+    delete m_images;
+}
+
+size_t OIIOImageFileWriter::get_image_count(void) const
+{
+    return m_images->m_canvas.size();
+}
+
+void OIIOImageFileWriter::append_image(const ICanvas* image)
+{
+    if (image == nullptr)
```

Comment:
An assertion would be more appropriate here: `image` being null is typically a programmer error.

Reply:
done